File: treatment/src/main/java/com/hartwig/actin/treatment/ctc/config/CTCDatabaseEntryFile.java
package com.hartwig.actin.treatment.ctc.config;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import com.hartwig.actin.util.ResourceFile;
import com.hartwig.actin.util.TabularFile;

import org.jetbrains.annotations.NotNull;

public final class CTCDatabaseEntryFile {

    private static final String DELIMITER = "\t";

    @NotNull
    public static List<CTCDatabaseEntry> read(@NotNull String tsv) throws IOException {
        List<String> lines = Files.readAllLines(new File(tsv).toPath());

        Map<String, Integer> fields = TabularFile.createFields(lines.get(0).split(DELIMITER));

        return lines.subList(1, lines.size()).stream().map(line -> create(fields, line.split(DELIMITER, -1))).collect(Collectors.toList());
    }

    @NotNull
    private static CTCDatabaseEntry create(@NotNull Map<String, Integer> fields, @NotNull String[] parts) {
        return ImmutableCTCDatabaseEntry.builder()
                .studyId(ResourceFile.integer(parts[fields.get("StudyID")]))
                .studyMETC(parts[fields.get("StudyMETC")])
                .studyAcronym(parts[fields.get("StudyAcroniem")])
                .studyTitle(parts[fields.get("StudyTitle")])
                .studyStatus(parts[fields.get("StudyStatus")])
                .cohortId(ResourceFile.optionalInteger(parts[fields.get("CohortId")]))
                .cohortParentId(ResourceFile.optionalInteger(parts[fields.get("CohortParentId")]))
                .cohortName(ResourceFile.optionalString(parts[fields.get("CohortName")]))
                .cohortStatus(ResourceFile.optionalString(parts[fields.get("CohortStatus")]))
                .cohortSlotsNumberAvailable(ResourceFile.optionalInteger(parts[fields.get("CohortSlotsNumberAvailable")]))
                .cohortSlotsDateAvailable(ResourceFile.optionalString(parts[fields.get("CohortSlotsDateAvailable")]))
                .build();
    }
}

File: treatment/src/main/kotlin/com/hartwig/actin/treatment/ctc/config/CTCDatabaseEntryFile.kt
package com.hartwig.actin.treatment.ctc.config

import com.hartwig.actin.treatment.FileUtil
import com.hartwig.actin.util.ResourceFile

object CTCDatabaseEntryFile {

    fun read(tsv: String): List<CTCDatabaseEntry> {
        return FileUtil.createObjectsFromTsv(tsv, ::create)
    }

    private fun create(fields: Map<String, Int>, parts: List<String>): CTCDatabaseEntry {
        return CTCDatabaseEntry(
            studyId = ResourceFile.integer(parts[fields["StudyID"]!!]),
            studyMETC = parts[fields["StudyMETC"]!!],
            studyAcronym = parts[fields["StudyAcroniem"]!!],
            studyTitle = parts[fields["StudyTitle"]!!],
            studyStatus = parts[fields["StudyStatus"]!!],
            cohortId = ResourceFile.optionalInteger(parts[fields["CohortId"]!!]),
            cohortParentId = ResourceFile.optionalInteger(parts[fields["CohortParentId"]!!]),
            cohortName = ResourceFile.optionalString(parts[fields["CohortName"]!!]),
            cohortStatus = ResourceFile.optionalString(parts[fields["CohortStatus"]!!]),
            cohortSlotsNumberAvailable = ResourceFile.optionalInteger(parts[fields["CohortSlotsNumberAvailable"]!!]),
            cohortSlotsDateUpdate = ResourceFile.optionalString(parts[fields["CohortSlotsDateUpdate"]!!])
        )
    }
}